packages/token-sdk/src/token-fetcher.ts

```diff
@@ -0,0 +1,105 @@
+import { Connection } from "@solana/web3.js";
+import { Address } from "@project-serum/anchor";
+import { Token } from "./types";
+import {
+  AddressUtil,
+  getMultipleParsedAccounts,
+  getParsedAccount,
+  ParsableMintInfo,
+} from "@orca-so/common-sdk";
+import { MintInfo } from "@solana/spl-token";
+import invariant from "tiny-invariant";
+import { MetadataProvider, MetadataUtil } from "./metadata";
+
+export class TokenFetcher {
+  private readonly connection: Connection;
+  private readonly _cache: Record<string, Token> = {};
+  private readonly providers: MetadataProvider[] = [];
+
+  private constructor(connection: Connection, cache: Record<string, Token> = {}) {
+    this.connection = connection;
+    this._cache = cache;
+  }
+
+  public static from(connection: Connection, cache?: Record<string, Token>): TokenFetcher {
+    return new TokenFetcher(connection, cache);
+  }
+
+  public addProvider(provider: MetadataProvider): TokenFetcher {
+    this.providers.push(provider);
+    return this;
+  }
+
+  public async find(address: Address): Promise<Token> {
+    const mint = AddressUtil.toPubKey(address);
+    const mintString = mint.toBase58();
+    if (!this._cache[mintString]) {
+      const mintInfo = await getParsedAccount(this.connection, mint, ParsableMintInfo);
+      invariant(mintInfo, "Mint info not found");
+      this._cache[mintString] = {
+        mint,
+        decimals: mintInfo.decimals,
+      };
+
+      for (const provider of this.providers) {
+        const metadata = await provider.find(mint);
+        const cachedValue = this._cache[mintString];
+        this._cache[mintString] = {
+          mint: cachedValue.mint,
+          decimals: cachedValue.decimals,
+          ...MetadataUtil.merge(cachedValue, metadata),
+        };
+        if (!MetadataUtil.isPartial(this._cache[mintString])) {
+          break;
+        }
+      }
+    }
+    return { ...this._cache[mintString] };
+  }
+
+  public async findMany(addresses: Address[]): Promise<Record<string, Token>> {
```

Comment:
nit: Promise<Readonly<Record<string, Token>> (or extract it to an exported type)

Also, do we care about TTL for tokens? should users have the power to force providers to always give the newest value?

Reply:
Added the `Readonly` types. Also removed creation of new objects when returning from the fetcher since the `Readonly` annotation protects modification of the returned cached values. Did a check and it gives a compile-time error.

Since the data should change very infrequently, my opinion is that we don't need to add a `refresh` capability right now. It should be easy to add in the future without changing the interface if we make the `refresh` param optional like in the `whirlpools-sdk`.